lib/ast_ml.ml

```diff
@@ -1,134 +1,95 @@
-      | ChoreoVar of {data: Var.t; position: Position.position}
-      (*Removing tycon: Tycon.t option because the type should be declared in Dec*)
-      | Expression of {data: (Location.t * LocalExpr.t); 
-          position: Position.position}
-      (*Removing tycon: Tycon.t option because the type should be declared in Dec*)
-      | Send of {data: sugaredSend; position: Position.position}
-      (*Removing tycon: Tycon.t option because the type should be declared in Dec*)
-      | Branch of {
-          data: (t * t * t); 
-          position: Position.position
-        }
-      | Sync of {
-          data: {
-            sender: Location.t;
-            receiver: Location.t;
-            label: Var.t;
-            then_: t
-          };
-          position: Position.position
-        }
-      | Let of {
-        data: {
-          declBlock: DeclBlock.t;
-          in_: t
-        },
-        position: Position.position
-      }
-      | Pair of {
-          data of t * t,
-          position: Position.position
-        }
-      | Fst of {
-        data of Pair.t,
-        position: Position.position
-      }
-      | Snd of {
-        data of Pair.t,
-        position: Position.position
-      }
-      | Match of {
-          data: {
-            matcher: t;
-            branches: (Pattern.t * t) list
-          },
-          position: Position.position
-        }
-      | Left of {
-        data of Match.t,
-        position: Position.position
+module rec S_P_TREE : sig
+  open Local_ast.L_P_TREE
+
+  type location = string
+
+  type tycon = string
+
+  and position = {
+    pos_fname : string;
+    pos_lnum : int;
+    pos_bol : int;
+    pos_cnum : int;
+  }
+
+  and pattern =
+    | Left of pattern
+    | Right of pattern
+    | Pattern of (location * lpattern)
+    | Pair of (pattern * pattern)
+    | Var of var
+    | Wild
+
+  and sugaredSend = 
+    | SendWithBinder of {
+        sender: location;
+        receiver: location;
+        binder: var;
+        expression: local_expr;
+        then_: choreography;
       }
-      | Right of {
-        data of Match.t,
-        position: Position.position
+    | SendWithoutBinder of {
+        sender: location;
+        receiver: location;
+        choreo: choreography;
       }
-  end
 
-  module Assignment : sig
-    type t =
-      | CType of {var: Var.t; tycon: Tycon.t}
-      | VarAsgn of {(Var.t * Tyvar.t option); location: Location.t option; choreo: Choreography.t}
-      | Fun of {var: Var.t; args: Pattern.t list; choreo: Choreography.t}
-  end
+  and syncData = {
+    sender: location;
+    receiver: location;
+    label: var;
+    then_: choreography;
+  }
+
+  and letData = {
+    declBlock: decl_block;
+    in_: choreography;
+  }
+
+  and matchData = {
+    matcher: choreography;
+    branches: (pattern * choreography) list;
+  }
+
+  and pair = (choreography * choreography)
+
+  and choreography =
```

Comment:
I might rename this to something like `chor`, since `choreography` is going to get really annoying to write really quickly.

Reply:
how about choreo?